src/middleware/persist.ts

```diff
@@ -0,0 +1,353 @@
+import { StateInitializer, Store, UnknownState } from "../vanilla"
+import { thenablify, Thenable, emitter } from "./utils"
+
+// ============================================================================
+// Types
+
+type Persist =
+  <T extends UnknownState, U>
+    ( stateInitializer: StateInitializer<T, Store<T>>
+    , options: PersistOptions<T, U>
+    ) =>
+      StateInitializer<T, Store<T> & PersistStore<T, U>>
+
+interface PersistOptions<T extends UnknownState, U>
+  { /** Name of the storage (must be unique) */
+    name: string
+    
+    /**
+     * A function returning a storage.
+     * The storage must fit `window.localStorage`'s api (or an async version of it).
+     * For example the storage could be `AsyncStorage` from React Native.
+     *
+     * @default () => localStorage
+     */
+  , getStorage?: () => PersistentStorage
+
+    /**
+     * Use a custom serializer.
+     * The returned string will be stored in the storage.
+     *
+     * @default JSON.stringify
+     */
+    serialize?:
+      (storageValue: PersistentStorageValue<U>) => MaybePromise<string>
+
+    /**
+     * Use a custom deserializer.
+     * Must return an object matching PersistentStorageValue<State>
+     *
+     * @param serializedString The storage's current value.
+     * @default JSON.parse
+     */
+    deserialize?:
+      (serializedString: string) => MaybePromise<PersistentStorageValue<T>>
+
+    /**
+     * Filter the persisted value.
+     *
+     * @param state The state's value
+     */
+    partialize?: (state: T) => U
```

Comment:
Previously this was `(state: T) => DeepPartial<T>` but I don't understand why was it restricted to `DeepPartial<T>`? `partialize` acts like a `selector`/`mapper` the user should be able to map to any shape they wish.